### ISSUE TITLE
None of the plugins are working since the switch to poetry

### ISSUE THREAD

**Issue by ClayShoaf** [ElementsProject/lightning#5115]
TL;DR How am I supposed to install and run plugins now that the project is using poetry? 

I've been working for months now on just trying to get to the point where I could actually contribute some small things to plugins that I thought would be helpful. I finally got to the point where I could actually build c-lightning. Took me forever to figure out that I had to do `poetry run make` .  Is there some new way that I'm supposed to be using plugins? 

At first I was trying to run tests by trying everything I could think of.  
I tried copying the plugin to the `lightning/plugins` directory and using `poetry run pytest` (both before and after hackily making sure I had the right dependencies) to test it. That didn't work.
I tried exporting the poetry dependencies and installing all of them using pip and running pytest natively. That didn't work.
Finally, I just tried to see if the plugin would actually work at all. When I used `contrib/startup_regtest.sh` and tried to start the plugin, I get the following error:

```
[user@prgrmr ~/github/lightning]$ l1-cli plugin start /home/user/github/lightning/plugins/noise/noise.py
Traceback (most recent call last):
  File "/home/user/github/lightning/plugins/noise/noise.py", line 5, in <module>
    from pyln.client import Plugin, RpcError
ModuleNotFoundError: No module named 'pyln'
{
   "code": -3,
   "message": "/home/user/github/lightning/plugins/noise/noise.py: exited before replying to getmanifest"
}
```
Probably has to do with the fact that I'm using pyenv and it's somehow trying to use my system version of python, idk, maybe it's something else. I can't even install any of the pyln stuff because it all used relative paths, and pip doesn't recognize things like `../` as a valid URL. I could copy the `pyln-*/pyln` folders manually to my site-packages folder like I did for the version I'm using in pyenv, but there is no way that this method is sustainable. There has to be something I'm missing. I'm almost at the end of my rope here.

Before this switch to poetry, I was just starting to get things to work on my Arch machine by using `pyenv` to use a compatible version of python. 
Tests were working and I was finally getting ready to work on the stuff that I actually wanted to work on. All I had to do was send a PR to add Arch installation instructions to `doc/INSTALL.md` and I would have been all set to contribute and actually write tests and everything. 

Now it's like I'm completely back to square one. 

Weeks worth of tweaking/testing and figuring things out was all for nothing. This is extremely frustrating.

**Comment by vincenzopalazzo** [comment]
>Probably has to do with the fact that I'm using pyenv and it's somehow trying to use my system version of python, idk, maybe it's something else. I can't even install any of the pyln stuff because it all used relative paths, and pip doesn't recognize things like ../ as a valid URL. I could copy the pyln-*/pyln folders manually to my site-packages folder like I did for the version I'm using in pyenv, but there is no way that this method is sustainable. There has to be something I'm missing. I'm almost at the end of my rope here.

Ok, I didn't reproduce in a clean machine the new dependencies manager. However, I think pip and poetry can work together by just running all in the same virtual env (or disabling the virtual env in poetry conf).

I suggested the following steps if you are building from source c-lightning and use the pyln somewhere else.

```
poetry config settings.virtualenvs.create false
poetry install
make && make install
l1-cli plugin start /home/user/github/lightning/plugins/noise/noise.py
```

The error should be gone, or at least different from what you see.

If you want build c-lightning to run the unit and integration testing, I suggest to use the following steps

```
poetry shell
poetry install
make && make check DEVELOPER=1 VALGRIND=0
```
This should be installed in a local env for all the dependencies and run c-lightning tests in this virtual env. This conf is better for development mode because you can have different dependencies.

In addition, I have no the opportunity to test in a clean machine, so any feedback are welcome :)

P.S: Python dependencies are always frustrating!

**Comment by ClayShoaf** [comment]
> Python dependencies are always frustrating!

Glad to know I'm not the only one.

I'll try out your suggestion this weekend. Thank you for the reply

**Comment by ClayShoaf** [comment]
Ok, I'm trying this on a fresh install of Ubuntu 20.04, and when I try `poetry config settings.virtualenvs.create false` I get an error message.

```
user@user-ThinkPad-T420:~/github/lightning$ poetry config settings.virtualenvs.create false

  ValueError

  Setting settings.virtualenvs.create does not exist

  at ~/.local/lib/python3.8/site-packages/poetry/console/commands/config.py:248 in handle
      244│                 raise ValueError("You must pass exactly 1 value")
      245│ 
      246│             return 0
      247│ 
    → 248│         raise ValueError("Setting {} does not exist".format(self.argument("key")))
      249│ 
      250│     def _handle_single_value(self, source, key, callbacks, values):
      251│         validator, normalizer, _ = callbacks
      252│ 
```
I went ahead and tried installing c-lightning as per `INSTALL.md` (with the slight change that I had to run `poetry run make` instead of just `make`. I'm getting the same error as in my original post when I try to run a plugin in regtest (using the `contrib/startup_regtest.sh` script. I haven't tried running anything outside of regtest because I don't have enough space on this computer to download the blockchain).

**Comment by vincenzopalazzo** [comment]
Sorry, the `settings.virtualenvs.create` is deprecated or removed.

can you try with `poetry config virtualenvs.create true`?

This PR should contains the correct changes to run poetry https://github.com/ElementsProject/lightning/pull/5133

**Comment by ClayShoaf** [comment]
I was able to install/run/test c-lighting with poetry. I can run `poetry config virtualenvs.create true` and verified with `poetry config --list`. Trying to build now and seeing what happens. Should I have tried it with "false" for the purposes of this testing?

Also, side note, what would be the point of using poetry if you're not going to use a virtual environment?

**Comment by ClayShoaf** [comment]
I'm not able to run plugins building with `poetry config virtualenvs.create true` or `poetry config virtualenvs.create false`. I tried running 
```
poetry install
./configure --enable-developer
poetry run make
sudo make install
```
after each change. Still getting the same error as in the original message.

**Comment by vincenzopalazzo** [comment]
> Also, side note, what would be the point of using poetry if you're not going to use a virtual environment?

dependencies resolution from the development side nothing related to the end-user!

>Trying to build now and seeing what happens. Should I have tried it with "false" for the purposes of this testing?

If you disable the env you are using the global env common to all python programs, this means that the plugin running the test version, and maybe it can be not you want.

I have a common machine for development, and when I develop on c-lightning I run `make check` inside the `poetry shell`

Python and pip are messy and poetry introduces some of the standards for running and managing pip in a good manner!

**Comment by ClayShoaf** [comment]
Trying this now inside of the poetry shell. Gonna take me a minute because I am `SLOW_MACHINE=1` (another setting that took me forever to find and was originally causing failed tests)

**Comment by vincenzopalazzo** [comment]
Can you please run the command in the correct sequence as described in the PR https://github.com/ElementsProject/lightning/pull/5133?

To summarize

```
poetry config virtualenvs.create false
poetry install
./configure --enable-developer
sudo make install
```

P.S: when you are running `sudo make install` you are changing your own user, this mean that you can have in your conf two different python env, one of your normal user, and another one for the sudo user

Some more content https://askubuntu.com/questions/234758/how-to-use-a-python-virtualenv-with-sudo

**Comment by ClayShoaf** [comment]
```
poetry config virtualenvs.create false
poetry install
./configure --enable-developer
sudo make install
```

That is the last one I tried. Still the same problem as the original post. I am trying to run everything inside of `poetry shell` now. If that doesn't work, I'm going to reinstall a fresh copy of Ubuntu and start from the beginning in case I accidentally messed with anything that might be causing problems. I wish I could just let you ssh into my machine so you could see. I'm surprised that no one else is having this problem.

**Comment by vincenzopalazzo** [comment]
Mh! I talk with someone about this problem too, on discord.

Before restarting from the clean machine, you could try also this?

```
poetry export -f requirements.txt --output requirements.txt
pip3 install -r requirements.txt
./configure --enable-developer
sudo make install
```

**Comment by ClayShoaf** [comment]
```
poetry export -f requirements.txt --output requirements.txt
pip3 install -r requirements.txt
./configure --enable-developer
sudo make install
```
That was one of the things I tried on a machine when I was getting frustrated. I believe that doesn't work because pyln uses relative paths `../../` which pip does not understand but poetry does.

However, after going through the whole process again in `poetry shell` IT WORKED! I am finally able to actually start a plugin as long as I do everything inside of a `poetry shell`. Now I just need to make sure I can actually run tests that pass.

If I am running inside of poetry shell, should I just use `pip` to install requirements-dev.txt for a plugin, or should I install each requirement using `poetry add`  or something like that from the plugin directory?

**Comment by ClayShoaf** [comment]
Thank you so much for your help @vincenzopalazzo . So the main thing that fixed it for me appears to be that I needed to run everything inside of `poetry shell`. It seems obvious now that I'm thinking about it (and, more importantly, now that I know that poetry shell exists, haha).

At some point I'll be able to send some sats to your ln address with a message attached, like the keysend billboard on [amboss.space](https://amboss.space) (which is the entire thing that is currently animating me right now).

I'm sorry that my submissions are probably sloppy and unorthodox. I'm kinda learning as I go in my limited free time. I know I'm basically the equivalent of some random carpenter trying to do electrical work on his house and asking questions to industrial/commercial electricians, and I know how puzzling it can be to see the ideas that people have sometimes when they don't understand the true nature of the underlying systems that they are trying to work with.

All of this is to say thank you for holding my hand through this process. Now I can at least take the next step forward and try to figure out why the tests are failing without taking up any more of your time. 

Maybe an edit to HACKING.md is needed to point out the use of `poetry shell`, but I'll save that for later when I'm more comfortable actually getting things running.

**Comment by vincenzopalazzo** [comment]
>All of this is to say thank you for holding my hand through this process. Now I can at least take the next step forward and try to figure out why the tests are failing without taking up any more of your time.

>Maybe an edit to HACKING.md is needed to point out the use of poetry shell, but I'll save that for later when I'm more comfortable actually getting things running.

Happy that you fixed the error, However, the poetry shell is not a solution to install the stuff globally. Thank you to point out this problem, I will look more deep inside and reporting a real solution for all

**Comment by ClayShoaf** [comment]
I'm just using it for regtest hacking at this point, so it will do fine for my purposes for now. I'll leave the issue open until there is a commited PR that you think satisfies the issue.

**Comment by vincenzopalazzo** [comment]
this issue can be close right @ClayShoaf ?

**Comment by ClayShoaf** [comment]
Yea. I suppose this could be considered resolved. 

If anyone is coming across this at some point in the future, you can output all of the requirements needed for plugins as a pip readable `requirements.txt` file by running
```
poetry export -f requirements.txt
```
You will have to manually edit the `requirements.txt` file to remove anything to do with `pyln` and then copy those things to your python library's `site-packages` folder from the `contrib/pyln-*/pyln/`

Maybe this should be included somewhere in `doc/INSTALL.md` but that whole doc needs an overhaul, from what I can tell. I haven't tried going through all of the different distro install instructions, but I would imagine that following them verbatim, as it is now, would not result in a functioning build of Core Lightning.

**Comment by SimonVrouwe** [comment]
>Also, side note, what would be the point of using poetry if you're not going to use a virtual environment?

@ClayShoaf Good point.
All plugins run in the same environment as `lightningd` runs in and that environment should thus satisfy the requirements of potentially **all** plugins! C-lightningd has no way to switch `venv` when starting a plugin, also not indirectly via a `poetry run <plugin>` call because its `pipecmd` only takes a single argument (the path-to-plugin-executable, see code below).
https://github.com/ElementsProject/lightning/blob/b6d334de1da9f573c2938a4388e60dc3b2ab6b7a/lightningd/plugin.c#L1735-L1739 


>dependencies resolution from the development side nothing related to the end-user!

@vincenzopalazzo That statement sounds a bit unfair, as some [README](https://github.com/lightningd/plugins/blob/master/backup/README.md#installation) explicitly instructs the end-user: "There are some Python dependencies. You can install them using poetry"

>Python and pip are messy and poetry introduces some of the standards for running and managing pip in a good manner!

@vincenzopalazzo That is a nice marketing slogan. But we have a bunch of separate plugins, with potentially conflicting requirements. I guess this hasn't been a problem so far because running a combination of plugins with conflicting requirements is rare.

So for now I recommend to _also_ keep `requirements.txt` available and up-to-date for every plugin, so that end-users can `pip3 install -r requirements.txt` inside c-lightning's common (virtual) environment or as its user (with `--user`).

Ideally plugins should be self-contained, have their own tests and their developers shouldn't be worried too much about being steam-rolled by generic solutions.

For the long term maybe it is possible to modify the `pipecmd` above to switch environment per-plugin, either by allowing extra args such as `poetry run <plugin>` or at a lower level by using `execvpe` (which takes an envp [argument](https://linux.die.net/man/3/execvpe))  instead of `execvp`. But I have no idea if that is feasible and how much overhead/complexity (`poetry run`) it adds.

**Comment by sha-265** [comment]
I have the same problem with CLN 23.08.1, any suggestions? I installed CLN with the binaries and not from source so I didn't use poetry. How can I solve this problem?
